src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java

```diff
@@ -552,103 +552,95 @@ private void preprocessACLs(final Query<Component> query, final String inputFilt
+            return dependencyGraph;
         }
         String queryUuid = ".*" + component.getUuid().toString() + ".*";
         final Query<Component> query = pm.newQuery(Component.class, "directDependencies.matches(:queryUuid) && project == :project");
         List<Component> components = (List<Component>) query.executeWithArray(queryUuid, project);
-        Map<String, Set<String>> pathComponentsMap = new HashMap<>();
         for (Component parentNodeComponent : components) {
-            if (pathComponentsMap.containsKey(parentNodeComponent.getUuid().toString())) {
-                pathComponentsMap.get(parentNodeComponent.getUuid().toString()).add(component.getUuid().toString());
+            parentNodeComponent.setExpandDependencyGraph(true);
+            if (dependencyGraph.containsKey(parentNodeComponent.getUuid().toString())) {
+                parentNodeComponent.getDependencyGraph().add(component.getUuid().toString());
             } else {
+                dependencyGraph.put(parentNodeComponent.getUuid().toString(), parentNodeComponent);
                 Set<String> set = new HashSet<>();
                 set.add(component.getUuid().toString());
-                pathComponentsMap.put(parentNodeComponent.getUuid().toString(), set);
+                parentNodeComponent.setDependencyGraph(set);
             }
-            getParentDependency(project, parentNodeComponent.getUuid().toString(), pathComponentsMap);
+            getParentDependenciesOfComponent(project, parentNodeComponent.getUuid().toString(), dependencyGraph);
         }
-        if (pathComponentsMap.isEmpty() && project.getDirectDependencies().contains(component.getUuid().toString())){
-            Set<String> set = new HashSet<>();
-            set.add(component.getUuid().toString());
-            pathComponentsMap.put(project.getUuid().toString(), set);
+        if (!dependencyGraph.isEmpty() || project.getDirectDependencies().contains(component.getUuid().toString())){
+            dependencyGraph.put(component.getUuid().toString(), component);
+            getRootDependencies(dependencyGraph, project);
+            getDirectDependenciesForPathDependencies(dependencyGraph);
         }
-        if (!pathComponentsMap.isEmpty()){
-            loadDependencyGraphPaths(pathComponentsMap, project, component.getUuid().toString());
+        // Reduce size of JSON response
+        for (Component c : dependencyGraph.values()) {
+            Component transientComponent = new Component();
+            transientComponent.setUuid(c.getUuid());
+            transientComponent.setName(c.getName());
+            transientComponent.setVersion(c.getVersion());
+            transientComponent.setPurlCoordinates(c.getPurlCoordinates());
+            transientComponent.setDependencyGraph(c.getDependencyGraph());
+            transientComponent.setExpandDependencyGraph(c.isExpandDependencyGraph());
+            dependencyGraph.put(c.getUuid().toString(), transientComponent);
```

Comment:
To my understanding this is not an issue, because you're replacing an existing entry, not adding or removing entries.

But maybe this could be made more clear by iterating over `entrySet` instead of `values`, and then using `dependencyGraph.put(entry.getKey(), transientComponent);`.